sympy/solvers/inequalities.py

```diff
@@ -982,3 +986,594 @@ def reduce_inequalities(inequalities, symbols=[]):
+            elif i.function == Q.eq:
+                i = Eq(i.rhs, i.lhs, evaluate=False)
+            else:
+                raise TypeError('only Ge, Le, or Eq is allowed, not %s' % i)
+        if i == True:
+            continue  # ignore
+        if i == False:
+            return  # no solution
+        if isinstance(i, Eq):  # could also collect k equalities and solve for k variables and eliminate them
+            L = i.lhs - i.rhs
+            np.extend([L, -L])
+        elif isinstance(i, (Le, Ge)):
+            i = i.canonical
+            npi = i.lts - i.gts
+            x = npi.free_symbols
+            if len(x) == 1:
+                x = x.pop()
+                if x in unbound:
+                    continue  # will handle later
+                ivl = Le(npi, 0, evaluate=False).as_set()
+                if x not in univariate:
+                    univariate[x] = ivl
+                else:
+                    univariate[x] &= ivl
+            else:
+                np.append(npi)
+        else:
+            raise TypeError('only Ge, Le, or Eq is allowed, not %s' % i)
+
+    # introduce auxilliary variables as needed for univariate
+    # inequalities
+    for x in univariate:
+        i = univariate[x]
+        if not i:
+            return None  # no solution possible
+        a, b = i.inf, i.sup
+        if a.is_infinite and b.is_infinite:
+            unbound.append(x)
+        elif a.is_infinite:
+            u = next(ui)
+            r[x] = b - u
+            aux.append(u)
+        elif b.is_infinite:
+            if a:
+                u = next(ui)
+                r[x] = a + u
+                aux.append(u)
+            else:
+                # standard nonnegative relationship
+                pass
+        else:
+            u = next(ui)
+            aux.append(u)
+            # shift so u = x - a => x = u + a
+            r[x] = u + a
+            # add constraint for u <= b - a
+            # since when u = b-a then x = u + a = b - a + a = b:
+            # the upper limit for x
+            np.append(u - (b - a))
+
+    # make change of variables for unbound variables
+    for x in unbound:
+        u = next(ui)
+        r[x] = u - x  # reusing x
+        aux.append(u)
+
+    return np, r, aux
+
+
+def _linprog_max(objective, constraints, unbound=None):
+    """return A, B, C, D, r, x+aux, aux for maximizing
+    objective = Cx - D with constraints Ax <= B, introducing
+    introducing aux variables as necessary to make replacements
+    of symbols as given in r, {sym: aux expression}, so all
+    variables in xx will take on nonnegative values.
+
+    If constraints contain symbols not in objective, this will
+    cause problems if passed to _simplex which expects all
+    elements to be Float or Rational.
+    """
+
+    # sympify input and collect free symbols
+    f = sympify(objective)
+    syms = list(ordered(f.free_symbols))
+    constraints = [sympify(i) for i in constraints]
```

Comment:
I don't think that this should use free symbols. For one thing it won't work with expressions like `f(x)` e.g.:
```python
linprog(f(x) + f(x).diff(x), [f(x) >= 0, f(x).diff(x) >= 0, f(x) <= 2*f(x).diff(x)])
```
The user should be expected to provide an ordered list of symbols if the input is not in the form of matrices. Mathematica's `LinearOptimisation` function has the signature `LinearOptimisation(objective, constraints, vars)` but can alternatively accept matrices like `LinearOptimisation(f_vec, (c_mat, b_vec))`.

Reply:
I don't think anything but symbols should be accepted since we are assuming a linear, not arbitrary, function.